Sources/MongoSwift/Operations/Operation.swift

```diff
@@ -38,3 +41,25 @@ internal func encodeOptions<T: Encodable>(options: T?, session: ClientSession?)
     try session?.append(to: &doc)
     return doc
 }
+
+/**
+ * This method takes in a client and an optional session and determines the appropriate connection the method should
+ * use to interact with mongoc.
+ * If session is non-nil: returns the session's underlying connection. Throws an error if the session is inactive or
+ * was not created from the provided client.
+ * If session is nil: returns a new connection from the client's underlying pool. **The caller is responsible for
+ * ensuring the returned connection is checked back into the pool.**
+ */
+internal func resolveConnection(client: MongoClient, session: ClientSession? = nil) throws -> Connection {
```

Comment:
The way this is currently written, it's impossible for the caller to know whether they should be returning the connection to the pool or not (i.e. whether or not the connection came from a session or from the pool). Above, we have to unwrap the session to figure this out (and pass it back in as an optional anyways). 

I think it could be useful to combine this function and `pool.withConnection` into a single `withResolvedConnection(client:session:func)` which would return the connection in the appropriate cases.

That way we could rewrite above as something like:
```swift
internal func execute<...>(...) {
    return try withResolvedConnection(client: client, session: session) {
        try operation.execute(using: $0, session: session)
    }
}
```
And have the guarantee that the connection is returned when it should be (and not when it shouldn't).

Reply:
ya, I felt a little weird about the fact that you can't tell unless if you inspect whether the session you passed in was `nil`.

your proposal works for the case above, and the case where we are re-doing the code from above in `bulkWrite`.

however, we are calling `resolveConnection` from a bunch of places that return a cursor, and in those cases we don't want to check the connection back in til later. (same will be true for change stream methods).

it's also kind of weird in the cursor case/change stream case, because we rely on the presence of `self.session` to determine whether we need to return the connection. maybe having `resolveConnection` be explicit and return both  a `Connection` and a boolean indicating whether the connection is now owned by the caller would be better?

we could then add `withResolvedConnection` which returns the connection to the pool if `resolveConnection`

and then maybe cursor/change streams could store an explicit boolean indicating whether they own their `Connection`.

Reply:
I was actually thinking about the `resolveConnection` duplication for cursors above. In node I chose to have cursors constructed with an operation, so they could reuse `executeOperation` internally. We have different requirements here, but could we perhaps accept e.g. a lambda in the `MongoCursor` constructor, so you'd have some pseudocode like:

```
 return try MongoCursor(from: cursor, client: self, connection: conn, decoder: self.decoder, session: session, initializer: { conn -> OpaquePointer in 
   // call your mongoc function here
})
```
(caveat emptor: I have no recollection what closure capture rules are in swift, this probably won't even compile)

The idea mostly would be that you could essentially treat the initializer like an operation, and localize the connection resolution to one place. This would have the added benefit of only having `resolveConnection` called in two places in the codebase, at which point the confusion Patrick points out might be more acceptable.

Reply:
Since we've already verified that the session exists and we therefore know the connection we're checking out will come from it and not the pool, I think it would be clearer if, instead of invoking `resolveConnection` here, we added some of the below logic to a method on `ClientSession` itself and called that. Then we'd reserve usage of `resolveConnection` exclusively for circumstances in which it's ambiguous where the connection could come from. 

This is a bit nitpicky I know, but imo it clears up the intention of the code a little bit.

Reply:
so I added back the initial `getConnection(forUseWith client: MongoClient)` method and now that's used instead.
then I realized at that point the only place we were even using `resolveConnection` anymore was in the `MongoCursor` initializer, now that we have this new lambda thing for cursor init. so I've just deleted the method and moved the logic there. 
I think that is clearer because `MongoCursor.init` now mirrors `MongoCursor.deinit` - we both assign and clean up `self._connection` based on whether `self._session` is nil.